catkit2/services/oceanoptics_spectrometer/oceanoptics_spectrometer.py

```diff
@@ -0,0 +1,187 @@
+        The model of the spectrometer.
+    pixels_number : int
+        number of pix in the spectras
+    wavelengths : NDArray[numpy.float_]
+        A NDArray with the wl of the spectrometer.
+    spectras : DataStream
+        A data stream to submit the spectras from the spectrometer.
+    should_be_acquiring : threading.Event
+        An event to signal whether the spectra should acquire a spectrum.
+    NUM_FRAMES : int
+        The number of frames to allocate for the data streams.
+
+    Methods
+    -------
+    open()
+        Open the service.
+    main()
+        The main function of the service.
+    close()
+        Close the service.
+    start_acquisition()
+        Start the acquisition.
+    take_one_spectrum()
+        Get a spectrum from the spectrometer.
+    exposure_time()
+        Set the exposure time of the spectrometer.
+
+    '''
+    NUM_FRAMES = 20
+
+    def __init__(self):
+        '''
+        Create a new OceanOpticsSpectrometer service.
+
+        '''
+        super().__init__('oceanoptics_spectrometer')
+        
+        self.serial_number = None
+        self.spectrometer = None
+        self.model = None
+        self.pixels_number = None
+
+        self.wavelengths = None
+        self.spectras = None
+
+        self._exposure_time = None
+
+        self.should_be_acquiring = threading.Event()
+        self.should_be_acquiring.set()
+
+    def open(self):
+        '''
+        Open the service.
+
+        This function is called when the service is opened.
+        It initializes the spectrometer and creates the data streams and properties.
+
+        Raises
+        ------
+        RuntimeError
+            If the spectro cannot be found.
+
+        '''
+
+        # Find the spectrometer
+        self.serial_number = str(self.config['serial_number'])
+        try:
+            self.spectrometer = spct.from_serial_number(self.serial_number)
+        except SeaBreezeError:
+            raise ImportError(f'OceanOptics: Could not find spectrometer with serial number {self.serial_number}')
+
+        # exctract attributes
+        self.model = self.spectrometer.model
+        self.pixels_number = self.spectrometer.pixels
+        self.wavelengths = self.spectrometer.wavelengths()
+
+        # Define and set defaut exposure time
+        self._exposure_time = self.config.get('exposure_time', 1000)
+        self.exposure_time(self._exposure_time)
+
+        # Create datastreams
+        # Use the full sensor size here to always allocate enough shared memory.
+        self.spectras = self.make_data_stream('spectras', 'float32', [self.pixels_number], self.NUM_FRAMES)
+
+    def main(self):
+        '''
+        The main function of the service.
+
+        This function is called when the service is started.
+        '''
+        while not self.should_shut_down:
+            if self.should_be_acquiring.wait(0.05):
+                self.take_one_spectrum()
+                self.should_be_acquiring.clear()
```

Comment:
It looks to me like this would take exactly one spectrum before stopping the acquisition by clearing the `should_be_acquiring` flag. Is this what you are going for? In this case, where you want to get a spectrum on-command, it might be better to use the variant that for example the DMs are using - you create a thread for a function `monitor_command`. This method monitors the "command" data stream (which needs to be created in the init) inside a while loop. In there, you have a try... except... that tries to get a new frame from the command data stream. If it does, it calls your function `take_one_spectrum()` (the equivalent to `update_dm()` in the DM service), if it doesn't, it just passes and waits until it checks the data stream again.

DM example here, and all the lines that follow:
https://github.com/spacetelescope/catkit2/blob/7663307222cee45655932ded1887ac10e8065d90/catkit2/services/bmc_dm/bmc_dm.py#L56

If you opt for this, you don't need to bother with any of the start acquisition / end acquisition stuff.

Reply:
Ok, I am not sure to understand. My goal was to give the mean to do both single and loop images, actually, because I am not sure about the use case of the spectro. One observation only could be accessed with spectrometer.take_one_spectrum() or a loop mode with start_acquisition/end_acquisition. Let's talk when you come back, it will be faster than exchangeing over DMs.